File: sample-kanban-node/app/routes/board-requests.js
/**
 * returns the configured requests to deliver the board pages and components
 * @param {{service}} options config object containing reference to the service
 * @returns {*&{components: {}, pages: {login(*): Promise<*>, board(*): Promise<void>, table(*): Promise<*>}}}
 */
export const boardRequests = ({ service }) => ({
  pages: {
    async board (ctx) {
      const model = {
        user: parseUser(ctx.cookies.get('x-user-info')),
        statuses: await service.listStatuses(),
        tasks: await service.listTasks()
      }
      await ctx.render('pages/board', model)
    },
    async login (ctx) {
      const model = {
        users: await service.listUsers()
      }
      await ctx.render('pages/login', model)
    },
    async table (ctx) {
      const model = {
        user: parseUser(ctx.cookies.get('x-user-info')),
        statuses: await service.listStatuses(),
        tasks: await service.listTasks()
      }
      await ctx.render('pages/table', model)
    }
  },
  components: {
    async addTask (ctx) {
      const { description, status } = ctx.request.body
      await service.addTask({ description, status_id: status })
      const model = {
        user: parseUser(ctx.cookies.get('x-user-info')),
        status: await service.findStatus(status),
        tasks: await service.listTasks()
      }
      await ctx.render('components/category-lanes', model)
    },
    async updateTask (ctx) {
      const { id } = ctx.request.params
      const { description, status } = ctx.request.body
      await service.updateTask({ id, description, status_id: status })
      const model = {
        user: parseUser(ctx.cookies.get('x-user-info')),
        status: await service.findStatus(status),
        task: await service.findTask(id)
      }
      await ctx.render('components/task-card', model)
    },
    async deleteTask (ctx) {
      const { id } = ctx.request.params
      const status = await service.findStatusByTaskId(id)
      await service.deleteTask(id)
      const model = {
        user: parseUser(ctx.cookies.get('x-user-info')),
        tasks: await service.listTasks(),
        status
      }
      await ctx.render('components/category-lanes', model)
    },
    async removePerson (ctx) {
      const { id, personId } = ctx.request.params
      await service.removePerson({ id, personId })
      const task = await service.findTask(id)
      const user = parseUser(ctx.cookies.get('x-user-info'))
      await ctx.render('components/task-members', { user, task })
    },
    async joinTask (ctx) {
      const { id } = ctx.request.params
      const user = parseUser(ctx.cookies.get('x-user-info'))
      await service.joinTask({ task_id: id, person_id: user.id })
      const task = await service.findTask(id)
      await ctx.render('components/task-members', { user, task })
    }
  }
})

/**
 * helper to crack open user
 *
 * @param cookie user encoded
 * @returns {{[p: string]: string} | null}
 */
const parseUser = cookie => {
  if (!cookie) return null
  const [kId, kName] = cookie.split('&')
  const kvId = kId.split('=')
  const kvName = kName.split('=')
  return { [kvId[0]]: kvId[1], [kvName[0]]: kvName[1] }
}

File: sample-kanban-go/app/requests/board-requests.go
package requests

import (
	"fmt"
	"github.com/labstack/echo/v4"
	"github.com/sombriks/sample-testcontainers/sample-kanban-go/app/models"
	"github.com/sombriks/sample-testcontainers/sample-kanban-go/app/services"
	"github.com/sombriks/sample-testcontainers/sample-kanban-go/app/templates/components"
	"github.com/sombriks/sample-testcontainers/sample-kanban-go/app/templates/pages"
	"log"
	"net/http"
)

type BoardRequest struct {
	service *services.BoardService
}

// NewBoardRequest - provision the request handlers for the kanban
func NewBoardRequest(service *services.BoardService) (*BoardRequest, error) {
	request := &BoardRequest{service: service}

	return request, nil
}

func getUser(c echo.Context) *models.Person {
	inContext := c.Get("user")
	if inContext == nil {
		return nil
	}
	return (inContext).(*models.Person)
}

// Index - simple redirect to board page
func (r *BoardRequest) Index(c echo.Context) error {
	return c.Redirect(302, "/board")
}

// BoardPage - route to provision and serve the kanban board page
func (r *BoardRequest) BoardPage(c echo.Context) error {
	user := getUser(c)
	statuses, err := r.service.ListStatus()
	if err != nil {
		return err
	}
	tasks, err := r.service.ListTasks("")
	if err != nil {
		return err
	}
	return pages.BoardPage(user, statuses, tasks).Render(c.Response().Writer)
}

func (r *BoardRequest) LoginPage(c echo.Context) error {
	user := getUser(c)
	people, err := r.service.ListPeople("")
	if err != nil {
		return err
	}
	return pages.Login(user, people).Render(c.Response().Writer)
}

func (r *BoardRequest) FakeLogin(c echo.Context) error {
	userId := c.FormValue("userId")
	var id int64
	_, _ = fmt.Sscan(userId, &id)
	user, _ := r.service.FindPerson(id)
	cookie := http.Cookie{
		Name:  "x-user-info",
		Value: user.UserToCookie(),
	}
	c.SetCookie(&cookie)
	return c.Redirect(302, "/board")
}

func (r *BoardRequest) FakeLogout(c echo.Context) error {
	cookie, err := c.Cookie("x-user-info")
	if err != nil {
		log.Println("[WARN] ", err.Error())
		return c.Redirect(302, "/login")
	}
	cookie.Value = ""
	c.SetCookie(cookie)
	return c.Redirect(302, "/login")
}

func (r *BoardRequest) CookieCheck(next echo.HandlerFunc) echo.HandlerFunc {
	return func(c echo.Context) error {
		cookie, err := c.Cookie("x-user-info")
		if err != nil {
			log.Println("[WARN] ", err.Error())
			return c.Redirect(302, "/login")
		}
		if cookie == nil || cookie.Value == "" {
			return c.Redirect(302, "/login")
		}
		c.Set("user", models.UserFromCookie(cookie.Value))
		return next(c)
	}
}

func (r *BoardRequest) TablePage(c echo.Context) error {
	user := getUser(c)
	return pages.TablePage(user).Render(c.Response().Writer)
}

func (r *BoardRequest) AddTask(c echo.Context) error {
	user := getUser(c)
	var statusId int64 = 0
	fmt.Sscan(c.FormValue("status"), &statusId)
	description := ""
	fmt.Sscan(c.FormValue("description"), &description)
	result, err := r.service.InsertTask(&models.Task{
		Description: description,
		StatusId:    statusId,
	})
	log.Printf("inserted task: %v\n", result)
	if err != nil {
		return err
	}
	status, err := r.service.FindStatus(statusId)
	if err != nil {
		return err
	}
	tasks, err := r.service.ListTasks("")
	if err != nil {
		return err
	}
	return components.CategoryLanes(user, status, tasks).Render(c.Response().Writer)
}

func (r *BoardRequest) UpdateTask(c echo.Context) error {
	user := getUser(c)
	var taskId int64 = 0
	var statusId int64 = 0
	fmt.Sscan(c.Param("id"), &taskId)
	fmt.Sscan(c.FormValue("status"), &statusId)
	_, err := r.service.UpdateTask(taskId, &models.Task{
		Description: c.FormValue("description"),
		StatusId:    statusId,
	})
	if err != nil {
		return err
	}
	status, err := r.service.FindStatus(statusId)
	if err != nil {
		return err
	}
	task, err := r.service.FindTask(taskId)
	if err != nil {
		return err
	}
	log.Printf("[INFO] updated task %v\n", task)
	return components.TaskCard(user, status, task).Render(c.Response().Writer)
}

func (r *BoardRequest) DeleteTask(c echo.Context) error {
	user := getUser(c)
	var taskId int64 = 0
	fmt.Sscan(c.Param("id"), &taskId)
	status, err := r.service.FindStatusByTaskId(taskId)
	if err != nil {
		return err
	}
	affected, err := r.service.DeleteTask(taskId)
	if err != nil {
		return err
	}
	tasks, err := r.service.ListTasks("")
	if err != nil {
		return err
	}
	log.Println("[INFO] ", affected, " rows affected")
	return components.CategoryLanes(user, status, tasks).Render(c.Response().Writer)
}

func (r *BoardRequest) RemovePerson(c echo.Context) error {
	user := getUser(c)
	var taskId int64 = 0
	fmt.Sscan(c.Param("id"), &taskId)
	var personId int64 = 0
	fmt.Sscan(c.Param("personId"), &personId)
	err := r.service.RemovePerson(taskId, personId)
	if err != nil {
		return err
	}
	task, err := r.service.FindTask(taskId)
	if err != nil {
		return err
	}
	return components.TaskMembers(user, task).Render(c.Response().Writer)
}

func (r *BoardRequest) JoinTask(c echo.Context) error {
	user := getUser(c)
	var taskId int64
	fmt.Sscan(c.Param("id"), &taskId)
	err := r.service.JoinTask(taskId, user.Id)
	if err != nil {
		return err
	}
	task, err := r.service.FindTask(taskId)
	if err != nil {
		return err
	}
	return components.TaskMembers(user, task).Render(c.Response().Writer)
}

func (r *BoardRequest) AddComment(c echo.Context) error {
	user := getUser(c)
	var taskId int64 = 0
	fmt.Sscan(c.Param("id"), &taskId)
	var content = c.FormValue("content")
	err := r.service.AddComment(taskId, user.Id, content)
	if err != nil {
		return err
	}
	task, err := r.service.FindTask(taskId)
	if err != nil {
		return err
	}
	return components.TaskComments(user, task).Render(c.Response().Writer)
}
